### PR TITLE
Inputs refactor and extra checking

### LINKED ISSUE #241: Refactor namelist inputs for coordinates and other inputs
Should we prioritise refactoring the coordinate, evolve_moments, drive, and remaining collisions inputs to use a namelist format?

This may have the benefit of reducing clutter in `moment_kinetics_input.jl`, as well an making the specification of the defaults unique. 

Is there any reason for why we should not refactor the input structure for these inputs?

@johnomotani @LucasMontoya4

### PR DESCRIPTION
**Pull request by johnomotani**
Organize the rest of the options that were still at the top level of the input file into sections. It would be easy to handle top-level options if we wanted them (I actually already wrote `set_defaults_and_check_top_level!()` for the `makie_post_processing` option reader), but I didn't see at the moment any options that it didn't make sense to put in some section.

All options are now declared using `set_defaults_and_check_section!()` so we check that each option that is given in the input exists. This PR also adds `check_sections!()` which checks that all the sections that are given in the input exist. This makes it impossible for a typo or out of date input file to mean an option that was meant to be set is not set - the 'worst' that could happen would be for a mistake to change the name of one option into a different option, but that should be pretty unlikely. At least nothing will be silently ignored now.

While modifying the 'options', takes the opportunity to remove the remaining `mutable struct` definitions in `input_structs.jl` to remove duplication (a couple of the `mutable struct`s were already not being used). The final options objects are still immutable, as they are (mostly) `NamedTuple`s created with `Dict_to_NamedTuple()`.

The new sections are:
* `[r]`, `[z]`, `[vperp]`, `[vpa]`, `[vzeta]`, `[vr]`, `[vz]` for the coordinates inputs
* `[em_fields]` - so far just for the `force_Er_zero_at_wall` option
* `[evolve_moments]` for the moment-kinetic flags, which are now `density`, `parallel_flow`, `parallel_pressure`, and `moments_conservation`
* `[reactions]` for the ion-neutral interaction parameters
* `[electron_fluid_collisions]` for the `nu_ei` parameter used by the `braginskii_fluid` electron model
* `run_name` and `base_directory` move into the `[output]` section

[update-input-file.jl.zip](https://github.com/user-attachments/files/16965125/update-input-file.jl.zip) is a (zip'ed) script that will convert old input files to the new format. You might want to run it to get a new input file, then copy over the new sections by hand to your original file, because the script will remove all comments and might reorder options and sections. It's currently written to rename the original file with a `.unmodified` suffix, but you could easily edit to add a suffix to the new filename instead if you prefer.

There are a couple of minor changes to defaults: the defaults for `vz` are no longer taken from the settings for `vpa`; the default for `neutral_numerical_dissipation:vz_dissipation_coefficient` is no longer taken from `ion_numerical_dissipation:vpa_dissipation_coefficient`. I don't think we often (ever?) relied on these defaults anyway, so hopefully this is not a disruptive change.

Closes #241.

**Edit 11/9/2024:** replaced the input file updater script with a fixed version. Thanks @LucasMontoya4 for noticing the bug!

### CODE REVIEW

**Comment by johnomotani** [comment]
PS this PR also adds a CI job that checks the scripts in `test_scripts/` run without errors.

**Comment by johnomotani** [comment]
This PR is pretty big because of all the changes to input and test files, but for the source code files in `moment_kinetics/src/` it reduces the number of lines of code by 424 (+718 -1142), so I can claim it simplifies things a bit!

File: moment_kinetics/test/gyroaverage_tests.jl [hunk 1]
```diff
@@ -107,14 +107,9 @@ function gyroaverage_test(absolute_error; rhostar=0.1, pitch=0.5, ngrid=5, kr=2,
         gyrophase_nelement_global = gyrophase_nelement_local # total number of elements 
         gyrophase_discretization = "finite_difference"
         gyrophase_L = 2.0*pi
+        gyrophase_bc = "periodic"
```
Comment (mrhardman) [inline moment_kinetics/test/gyroaverage_tests.jl:110]: Was adding this boundary condition string necessary? I don't think any calculus functions are used for this coordinate.

File: moment_kinetics/test/fokker_planck_time_evolution_tests.jl [hunk 2]
```diff
@@ -287,20 +288,24 @@ function run_test(test_input, expected, rtol, atol, upar_rtol=nothing; args...)
     end
 
     # Convert keyword arguments to a unique name
-    name = input["run_name"]
+    function stringify_arg(key, value)
```
Comment (mrhardman) [inline moment_kinetics/test/fokker_planck_time_evolution_tests.jl:291]: Should this function be in this test? Why not in utils.jl?

File: moment_kinetics/src/coordinates.jl [hunk 3]
```diff
@@ -129,40 +129,145 @@ struct coordinate{T <: AbstractVector{mk_float}}
 end
 
 """
+    get_coordinate_input(input_dict, name)
+
+Read the input for coordinate `name` from `input_dict`, setting defaults, etc.
+"""
+function get_coordinate_input(input_dict, name; ignore_MPI=false)
```
Comment (mrhardman) [inline moment_kinetics/src/coordinates.jl:136]: Why do we have to supply name and input_dict as separate options? Maybe it is obvious. I think name would have previously been inside the grid_input struct, and we would have passed just the inputs for the specific coordinate of interest, not the input_dict for all coordinates (and then slice with name).

File: moment_kinetics/test/calculus_tests.jl [hunk 4]
```diff
@@ -29,24 +28,25 @@ function runtests()
...
+                                                           "discretization"=>discretization,
+                                                           "finite_difference_option"=>fd_option,
+                                                           "cheb_option"=>cheb_option,
+                                                           "bc"=>bc,
+                                                           "element_spacing_option"=>element_spacing_option))
                 # create the coordinate struct 'x'
                 # This test runs effectively in serial, so use `ignore_MPI=true` to avoid
                 # errors due to communicators not being fully set up.
-                x, spectral = define_coordinate(input; ignore_MPI=true)
+                x, spectral = define_coordinate(input, "coord"; ignore_MPI=true, collision_operator_dim=false)
```
Comment (mrhardman) [inline moment_kinetics/test/calculus_tests.jl:49]: As an example of my question about this syntax, I don't see why we shouldn't have 

```
input = OptionsDict("name" => "coord", "ngrid" => ... etc)
x, spectral = define_coordinate(input; ignore_MPI=true)
```
Why do we ever need to pass the input for multiple coordinates to define_coordinate?

**Review by mrhardman — COMMENTED** [review]
Thanks for making this huge effort! I have a few questions about syntax that I'd like to discuss before going forward. My two big questions are "why remove input structs?" and "why pass input data of multiple coordinates to define_coordinate?".

File: moment_kinetics/src/gauss_legendre.jl
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/gauss_legendre.jl:1]: The refactor of the coordinates fixed a bug in the test. The test was setting `nrank=0`, which is nonsense, and so was following a strange code path. By a fluke that code path gave accurate results for derivatives. Fixing that bug made it apparent that periodic boundary conditions in `gauss_legendre` weren't actually working properly (e.g. the mass matrix was singular, so couldn't be inverted), so the changes were needed to make the test pass again.

File: moment_kinetics/src/input_structs.jl
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/input_structs.jl:1]: Not necessarily, just that names should be defined in as few places as possible (preferably one). One way to achieve this is to define the options and defaults with keyword arguments to `set_defaults_and_check_section!()`, and then convert the output to a `NamedTuple` to get an immutable struct that can be passed around (but whose names are implicitly defined when it's created instead of needing to be declared separately).

It is nice to be able to declare a `struct` explicitly, often this is clearer and in some ways easier to find, so I've just been writing a new version of `set_defaults_and_check_section!()` that can take an `@kwdef struct` and get the option names and default values from that struct definition. That's also quite a nice solution - its restriction is that there's not an easy way to be able to modify option values depending on other things (at least I can't see an easy-to-implement one at the moment), so it's only useful when the defaults are just constant values. It seems to be quite nice for the initial conditions inputs though - I can remove another 90 lines of code by using it there.

File: moment_kinetics/src/gauss_legendre.jl
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/gauss_legendre.jl:1]: This is unfortunate -- I suppose that I never properly tested the periodic option. Didn't we find when looking at the time advance in the collision operator that boundary conditions should not be imposed in the mass matrix? (Only in the stiffness matrices, if solving an ODE). I would be careful before adding boundary conditions to mass matrices again. Can we schedule a discussion on this?

File: moment_kinetics/src/input_structs.jl
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/input_structs.jl:1]: Ah, so is a NamedTuple immutable after creation?

Yes, when using the `@kwdef struct` for geometry and Fokker-Planck collisions I came up against the same limitation, and ended up with the existing code structure.

File: moment_kinetics/src/load_data.jl [hunk 5]
```diff
@@ -246,7 +246,9 @@ function load_coordinate_data(fid, name; printout=false, irank=nothing, nrank=no
         return nothing, nothing, nothing
     end
 
+    input = OptionsDict()
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/load_data.jl:249]: I'm not sure I'm following your concern. Within `define_coordinate()`, the `input` is passed through `set_defaults_and_check_section()`, so any parameter that wouldn't be used would cause an error, and any parameter that is not set will use its default value. It is possible that `load_coordinate_data()` could be made simpler and/or more robust, I haven't given it any thought yet (and am not planning to any time soon).

File: moment_kinetics/src/numerical_dissipation.jl [hunk 6]
```diff
@@ -41,65 +42,54 @@ vz_dissipation_coefficient
...
+    neutral_params = Dict_to_NamedTuple(neutral_settings)
 
-    ion_input_dict = Dict(Symbol(k)=>v for (k,v) in ion_input)
-    ion_params = ion_num_diss_params(; ion_input_dict...)
-    electron_input_dict = Dict(Symbol(k)=>v for (k,v) in electron_input)
-    electron_params = electron_num_diss_params(; electron_input_dict...)
-    neutral_input_dict = Dict(Symbol(k)=>v for (k,v) in neutral_input)
-    neutral_params = neutral_num_diss_params(; neutral_input_dict...)
-
-    return numerical_dissipation_parameters(ion_params, electron_params, neutral_params)
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/numerical_dissipation.jl:102]: `NamedTuple` is immutable.

I want to use `set_defaults_and_check_section!()` to read the inputs, because that's the function that registers the section as being 'expected' so we can check that only expected sections exist. Also, `set_defaults_and_check_section!()` has nicer error messages.

File: moment_kinetics/test/gyroaverage_tests.jl (hunk 1 above)
Comment (johnomotani) [reply moment_kinetics/test/gyroaverage_tests.jl:110]: Periodic is what was being set before (as a hard-coded argument to `grid_input`), so I didn't change it. It seemed more in keeping with the style of code in this function to include it like this.

File: moment_kinetics/test/fokker_planck_time_evolution_tests.jl (hunk 2 above)
Comment (johnomotani) [reply moment_kinetics/test/fokker_planck_time_evolution_tests.jl:291]: Ah, it could be now that it's used in a few places. Some of the tests have slightly different hacked versions to get short enough names to not cause crashes when passed to HDF5. You're welcome to consolidate them if it bothers you.

File: moment_kinetics/src/coordinates.jl (hunk 3 above)
Comment (johnomotani) [reply moment_kinetics/src/coordinates.jl:136]: `set_defaults_and_get_section!()` is called inside `get_coordinate_input()`, and those are the arguments it expects. `set_defaults_and_get_section!()` does actually require `input_dict`, so it can save the name of the section as being 'expected'.

File: moment_kinetics/test/calculus_tests.jl (hunk 4 above)
Comment (johnomotani) [reply moment_kinetics/test/calculus_tests.jl:49]: As above, so the code can check that only the right sections exist.

File: moment_kinetics/src/load_data.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/load_data.jl:249]: My concern is that one could imagine adding an extra parameter to the input parameter deck inside the call to `set_defaults_and_check_section()`, but not add the appropriate `input["new_parameter"] = ` here in load_data, and thus only every get the default value for "new_parameter", without getting an error. If we had to pass "new_parameter" to a struct first, we would get an error if we forgot to do it.

File: moment_kinetics/src/coordinates.jl [hunk 7]
```diff
@@ -129,40 +129,145 @@ struct coordinate{T <: AbstractVector{mk_float}}
...
+    if coord_input_dict["nelement_local"] == -1 || ignore_MPI
+        coord_input_dict["nelement_local"] = coord_input_dict["nelement"]
+    end
+    if name == "vperp" && coord_input_dict["bc"] == "default"
+        if coord_input_dict["ngrid"] == 1 && coord_input_dict["nelement"] == 1
+            # 1V simulation, so boundary condition should be "none"
+            coord_input_dict["bc"] = "none"
+        else
+            # 2V simulation, so boundary condition should be "zero"
+            coord_input_dict["bc"] = "zero"
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/coordinates.jl:189]: I think I was just trying to be lazy. Will look again...

File: moment_kinetics/src/numerical_dissipation.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/numerical_dissipation.jl:102]: I have nothing against `set_defaults_and_check_section!()`. My concern here was just about the object returned by the function. For geometry and composition and etc, we use explicitly defined structs. I find these easier to understand because the typing documentation is explicitly given somewhere. Happy to hear that NamedTuple is immutable, would be even happier if there was an explicit type definition of everything that goes into the NamedTuple (i.e., it was a struct : ) ). This is probably just a preference based on the other existing namelist options.

File: moment_kinetics/test/fokker_planck_time_evolution_tests.jl (hunk 2 above)
Comment (mrhardman) [reply moment_kinetics/test/fokker_planck_time_evolution_tests.jl:291]: This seems reason enough to leave it alone : )

File: moment_kinetics/src/coordinates.jl (hunk 3 above)
Comment (mrhardman) [reply moment_kinetics/src/coordinates.jl:136]: So this is just because you wanted to use an existing function? To me it seems that `set_defaults_and_get_section!()` doesn't have exactly the right functionality for the task here (we should be passing in inputs for a single coordinate, we shouldn't need to check that the data exists, we are passing it!).

File: moment_kinetics/test/calculus_tests.jl (hunk 4 above)
Comment (mrhardman) [reply moment_kinetics/test/calculus_tests.jl:49]: Why wouldn't the section exist if we are passing a unique set of coordinate data? I don't see that `set_defaults_and_get_section!()` is appropriate here, but I respect the choice to use existing functions.

File: moment_kinetics/src/coordinates.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/coordinates.jl:189]: To be fair, this is new information that was not in the source. The Fokker-Planck time evolution tests are the place to look for evidence of my claim.

File: moment_kinetics/src/coordinates.jl [hunk 8]
```diff
@@ -129,40 +129,145 @@ struct coordinate{T <: AbstractVector{mk_float}}
...
+    get_coordinate_input(input_dict, name)
+
+Read the input for coordinate `name` from `input_dict`, setting defaults, etc.
+"""
+function get_coordinate_input(input_dict, name; ignore_MPI=false)
+    if name == "z"
+        default_bc = "wall"
+    elseif name == "r"
+        default_bc = "periodic"
+    elseif name == "vperp"
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/coordinates.jl:141]: My feeling is that making the defaults too convoluted is probably more confusing than helpful. To have the defaults you suggest, the `bc` for vpa and vperp depends on both the Fokker-Planck collisions and the spatial coordinates (and potentially also the numerical dissipation). We could pass all that information into `define_coordinate()` for vpa and vperp, but I'm not sure it's worth bothering. In the interest of simplicity I'd lean towards saying that the default mode for the code is that spatial variation exists, so "zero" is a sensible default. If someone wants to run 0D Fokker-Planck simulations you just have to know to set the right boundary conditions. Don't we normally we have the "bc" set in the input file anyway?

File: moment_kinetics/test/calculus_tests.jl (hunk 4 above)
Comment (johnomotani) [reply moment_kinetics/test/calculus_tests.jl:49]: The idea is to catch mistakes in the input file, e.g. if you made a copy-paste-edit error and created a section called `[vpz]` when you meant `[vz]`. Since (with this PR's code) we know what sections are supposed to exist, we can throw an error and say that `[vpz]` is incorrect. This means that something global needs to know what sections are 'supposed to exist' - that information is stored in `input_dict`, and the fact that the section is 'supposed to exist' is declared by calling `set_defaults_and_check_section!()` (within `define_coordinate()`.

File: moment_kinetics/src/numerical_dissipation.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/numerical_dissipation.jl:102]: Yeah, there's not quite an ideal solution. I'd like to only 'declare' the options in one place - either as the fields of a struct or as the arguments to `set_defaults_and_check_section!()`. The trouble is that: if you declare the fields and the defaults in a struct definition, then the defaults have to be constants; if you use `get_defaults_and_check_section!()` to get an `OptionsDict` with the right variables, you'd have to declare them twice to have a struct that contains the same variables.

Using `NamedTuple` is a compromise where you do the second option, but have to live with having an implicitly-defined `NamedTuple` rather than an explicitly-defined struct. Sometimes having the names declared twice so you can have the struct will be the best thing to do.

See the new definition of initial conditions in 3ad41d6 for an example of the 'first option'.

File: moment_kinetics/test/calculus_tests.jl (hunk 4 above)
Comment (mrhardman) [reply moment_kinetics/test/calculus_tests.jl:49]: I see why this functionality is very very useful when checking input files, but not when checking the functionality for a specific `define_coordinate()` call. To me it just seems that we are reusing an i/o function to manage passing of data internally, with consequences that we need to pass name information twice. If I understand correctly, the check is done on the duplicated name information that is given in almost the same place as the data is defined, not against some global list of allowed coordinates.

File: moment_kinetics/src/numerical_dissipation.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/numerical_dissipation.jl:102]: I acknowledge the two options, and suggest a third -- my preference is for an option always involving an explicitly defined struct, at the expense of declaring the variables twice to get the right defaults, where necessary.

File: moment_kinetics/src/gauss_legendre.jl
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/gauss_legendre.jl:1]: I don't know what's the right thing to do. At least in the test, we're only looking at a second derivative of a periodic function in a periodic box, so we do know what the right answer is. The original version (if you actually gave consistent inputs that go into the 'periodic bc' branches), gave a singular mass matrix, so we just get an error. The updated version does give the right answer - that's my only argument that it' the thing to do. Given that we have the duplicate point on the grid, modifying the matrix to impose that those two points have an equal value is the only way to make the 'periodic' boundary the same as all the other element boundaries (where continuity is imposed by construction), so it seems like a not-unreasonable thing to do. I'd be happy to believe that periodic 'boundary conditions' are special in some sense (that I don't understand...) because they're not 'really' boundary conditions...

[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/gauss_legendre.jl:1]: These are fair points. I think the right thing to do is (i) have a call with a mathematics collaborator to see what they do (ii) make appropriate documentations saying that this might be a kludge and (iii) make a strong note telling developers not to add any other BCs to mass matrices (periodic BC is a special exception).

File: moment_kinetics/test/calculus_tests.jl (hunk 4 above)
Comment (johnomotani) [reply moment_kinetics/test/calculus_tests.jl:49]: No, we create a global list, but it's created implicitly (by the `set_defaults_and_check_section!()` calls), so that we don't have to maintain a global list separately, which is required to be consistent and would cause bugs when it isn't.

The calls to `define_coordinate()` in `mk_input()` add the `r`, `z`, `vperp`, `vpa`, `vzeta`, `vr`, `vz` sections to `input_dict["_section_check_store"]`. When all the input sections have been read, `input_dict["_section_check_store"]` contains the list of all the sections, and then we check in `check_sections!()` that the input didn't have any extra sections that weren't read.

So I think the current version is useful and necessary (to check that there are no extra sections that would be silently ignored). We could have an extra version of `define_coordinate()` that just took the input for a single coordinate with no name, and I get the point it would simplify some of the test code by not having to create a subsection to pass input, but then the tests wouldn't be using the same function as the main code, which would probably be OK but contains some risk of causing problems. I don't think getting rid of a few `"z" => OptionsDict(` characters is worth that (admittedly small) risk, and the question for a new user of 'why do calls to `define_coordinate()` in the tests look different from calls to `define_coordinate()` in `mk_input()?`.

Comment (johnomotani) [reply moment_kinetics/test/calculus_tests.jl:49]: Although if something like
```julia
function define_test_coordinate(input_dict)
    input_dict = deepcopy(input_dict)
    name = pop!(input_dict, "name")
    return define_coordinate(OptionsDict(name => input_dict), name)
```
would make you happy, we can add that. Could be worth doing, as it would simplify the syntax in some tests.

We could even have a kwargs version so we don't have to define an `OptionsDict` like
```julia
function define_test_coordinate(; kwargs...)
    return define_test_coordinate(OptionsDict(String(k) => v for (k,v) in kwargs)
end
```

Comment (johnomotani) [reply moment_kinetics/test/calculus_tests.jl:49]: Do you ever want distributed-MPI in the test coordinates? If so we'd need to pass some of the keyword arguments through `define_test_coordinate()`.

Comment (mrhardman) [reply moment_kinetics/test/calculus_tests.jl:49]: So far I haven't needed distributed MPI in the tests. Maybe that would be desirable in the future -- but I think the bigger challenge connected to that is making a distributed LU decomposition and distribution LU solve so that the weak-form calculus can actually use distributed MPI.

File: moment_kinetics/src/gauss_legendre.jl
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/gauss_legendre.jl:1]: Resolved by #249.

File: moment_kinetics/src/load_data.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/load_data.jl:249]: I think the chances of making mistakes actually go down with the new approach.

With the old approach, `grid_input` was created with positional arguments, which are prone to errors (e.g. being entered in the wrong order) - that problem could be fixed using `@kwdef`, but actually you can get that effect with the new structure if you want, by creating a `NamedTuple` of inputs manually, which would cause errors if any input was missing. I'm not convinced that's useful though, unless you object to having default values at all.

In the new approach, if the inputs change then either:
* an input is removed, and the test uses that input - in this case, there will be an error message saying that input does not exist if it is not updated.
* an input is renamed - again you get a helpful error
* a new input is added - existing tests should not depend on this input, and should keep working using its default value. If the default is not OK, then the tests should fail unless we are very unlucky, and if you just added a new input parameter it shouldn't be too hard to realise that the new parameter needs to be set.

File: moment_kinetics/src/numerical_dissipation.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/numerical_dissipation.jl:102]: That is an option, and we do use it in places. I find the `NamedTuple` clear enough (as the usage is pretty standard, the fields are defined in a call to `set_defaults_and_check_section!()` in a 'setup' function for some module) that it's worth using (sometimes) for the benefit of being easier to update.

I don't think we're going to agree on this. @mabarnes @LucasMontoya4 do you have an opinion? I think to sum up, the question is:
In the case where we want the default for one or more options to not just be a constant value (i.e. depend on another option, or some other parameters) there are two options
1. Define a struct for settings (using `@kwdef`); call `set_defaults_and_check_section!()` with the keyword arguments being the fields of the struct, creating an `OptionsDict`; do some processing to update the default values as necessary, modifying the `OptionsDict`; construct an instance of the struct using the `OptionsDict`.
2. Call `set_defaults_and_check_section!()` with the keyword arguments being the fields of the struct, creating an `OptionsDict`; do some processing to update the default values as necessary, modifying the `OptionsDict`; use `Dict_to_NamedTuple` to create a `NamedTuple` (which is an immutable, implicitly defined struct) from the `OptionsDict`.
1 has an explicitly declared struct, but requires all the fields to be 'defined' twice (once in the struct definition and once in the call to `set_defaults_and_check_section!()`. 2 defines the fields only once, but creates a `NamedTuple` with those fields.

Should we require 1 or 2, or leave it up to whoever is writing the module to decide which seems more convenient?

PS in this particular instance, we could avoid this issue by having constant defaults if we didn't bother to replace `-Inf` with `nothing` for `force_minimum_pdf_value`. I think using `nothing` is marginally more efficient because the `if` can be evaluated at compile-time, but the cost will be totally negligible anyway because it's one float comparison that's outside any loop over arrays. Maybe that would be the best option in this instance!

File: moment_kinetics/test/calculus_tests.jl (hunk 4 above)
Comment (johnomotani) [reply moment_kinetics/test/calculus_tests.jl]: I've added `define_test_coordinate()` now, so I think this is resolved.

File: moment_kinetics/src/load_data.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrhardman) [reply moment_kinetics/src/load_data.jl:259]: OK, these seem like convincing arguments.

File: moment_kinetics/src/numerical_dissipation.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnomotani) [reply moment_kinetics/src/numerical_dissipation.jl:102]: Actually, maybe this is not such a big issue after all. I got into the habit of using `NamedTuple`s when `set_default_and_check_section!()` couldn't handle structs at all, but this discussion with @mrhardman prompteb me to add the version that can directly create an `@kwdef struct` as long as the defaults are constant. Now (with the commits I just pushed) using that version where possible, there are only a few places where we create settings in a `NamedTuple` and:
* `evolve_moments` settings don't propagate far, only until they get put into the `moments` struct.
* output from `get_coordinate_input()` only gets used inside or fed into `define_coordinate()`.
* input for manufactured solutions is only used within that module, and it's useful to make adding new inputs as simple as possible there.
* the `nonlinear_solvers` functionality is still WIP, so it's useful to be able to change it simply (at least for now).
* The reference parameters aren't used in many places, and we might want to add more derived parameters.
